analytics-integrations/kahuna/src/main/java/com/segment/analytics/internal/integrations/KahunaEcommerceHelper.java

```diff
@@ -0,0 +1,115 @@
+    String discountString = track.properties().getString(SEG_DISCOUNT_KEY);
+    if (!TextUtils.isEmpty(discountString)) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PURCHASE_DISCOUNT, discountString);
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    } else {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PURCHASE_DISCOUNT, "0");
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processAddedProductCategory(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().category())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_ADDED_TO_CART_CATEGORY, track.properties().category());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processAddedProductName(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().name())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_ADDED_TO_CART_NAME, track.properties().name());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processViewedProductName(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().name())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_VIEWED_NAME, track.properties().name());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processProductCategory(TrackPayload track) {
+
+    if (!TextUtils.isEmpty(track.properties().category())) {
+      Set<String> kahunaCategoriesSet
+              = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
+        protected boolean removeEldestEntry(Entry<String, Boolean> eldest) {
+          return size() > MAX_CATEGORIES_VIEWED_ENTRIES;
+        }
+      });
+      if (KahunaAnalytics.getUserAttributes().containsKey(CATEGORIES_VIEWED)) {
+        String serializedCategories = KahunaAnalytics.getUserAttributes().get(CATEGORIES_VIEWED);
+        kahunaCategoriesSet.addAll(Arrays.asList(serializedCategories.split(",")));
+      }
+      kahunaCategoriesSet.add(track.properties().category());
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(CATEGORIES_VIEWED, TextUtils.join(",", kahunaCategoriesSet));
+      userAttributes.put(LAST_VIEWED_CATEGORY, track.properties().category());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    } else {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_VIEWED_CATEGORY, NONE);
+      if (userAttributes.get(CATEGORIES_VIEWED) == null) {
+        userAttributes.put(CATEGORIES_VIEWED, NONE);
```

Comment:
We track the category NONE because it is an optional field when tracking something like Last Product Viewed and we don't want our attributes to be out of sync by not tracking a last category correctly.

To answer your 2nd question, yes. Kahuna tracks Users and not devices, so from a user perspective, your last category viewed in your profile would correspond to the last category you actually looked at, regardless of phone, tablet or web. We handle merging as needed on our server side.